File: spark-scala/spark-rdd/spark-rdd-broadcast/bin/main/com/mahfooz/spark/rdd/broadcast/RDDBroadcast.scala
/*
Broadcast variables are shared variables across all executors.
Broadcast variables are created once in the Driver and then are read only on executors.
Entire datasets can be broadcast in a Spark cluster so that executors have access to the broadcast data.
All the tasks running within an executor all have access to the broadcast variables.
Broadcast uses various optimized methods to make the broadcast data accessible to all executors.
The driver can only broadcast the data it has and you cannot broadcast RDDs by using references.
This is because only Driver knows how to interpret RDDs and executors only know the particular partitions of
data they are handling.

Broadcast variables can be both created and destroyed too.

 */
package com.mahfooz.spark.rdd.broadcast

import org.apache.spark.sql.SparkSession

object RDDBroadcast {

  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder()
      .appName(RDDBroadcast.getClass.getName)
      .master("local[3]")
      .getOrCreate()

    val states =
      Map(("NY", "New York"), ("CA", "California"), ("FL", "Florida"))
    val countries = Map(("USA", "United States of America"), ("IN", "India"))

    val broadcastStates = spark.sparkContext.broadcast(states)
    val broadcastCountries = spark.sparkContext.broadcast(countries)

    val data = Seq(
      ("James", "Smith", "USA", "CA"),
      ("Michael", "Rose", "USA", "NY"),
      ("Robert", "Williams", "USA", "CA"),
      ("Maria", "Jones", "USA", "FL")
    )

    val rdd = spark.sparkContext.parallelize(data)

    val rdd2 = rdd.map(f => {
      val country = f._3
      val state = f._4
      val fullCountry = broadcastCountries.value.get(country).get
      val fullState = broadcastStates.value.get(state).get
      (f._1, f._2, fullCountry, fullState)
    })

    println(rdd2.collect().mkString("\n"))
  }
}

File: spark-python/pyspark-rdd/src/broacasting/rdd_broadcast.py
from pyspark.sql import SparkSession


def main():
    spark = SparkSession.builder.getOrCreate()
    student_data = [
        ("Chris", 1523, 0.72, "CA"),
        ("Jake", 1555, 0.83, "NY"),
        ("Cody", 1439, 0.92, "CA"),
        ("Lisa", 1442, 0.81, "FL"),
        ("Daniel", 1600, 0.88, "TX"),
        ("Kelvin", 1382, 0.99, "FL"),
        ("Nancy", 1442, 0.74, "TX"),
        ("Pavel", 1599, 0.82, "NY"),
        ("Josh", 1482, 0.78, "CA"),
        ("Cynthia", 1582, 0.94, "CA"),
    ]
    student_rdd = spark.sparkContext.parallelize(student_data)
    rdd_transformation = student_rdd.map(lambda x: (x[0], x[1], int(x[2] * 100), x[3]))

    states = {"NY": "New York", "CA": "California", "TX": "Texas", "FL": "Florida"}

    broadcastStates = spark.sparkContext.broadcast(states)

    rdd_broadcast = rdd_transformation.map(
        lambda x: (x[0], x[1], x[2], broadcastStates.value.get(x[3], x[3]))
    )

    # confirm transformation is correct
    rdd_broadcast.collect()
    for record in rdd_broadcast.collect():
        print(record)


if __name__ == "__main__":
    main()
